# O problema da dieta Stigler

Nesta aula, veremos como resolver um problema clássico chamado dieta Stigler, nomeado em homenagem ao prêmio Nobel de Economia George Stigler, que calculou uma maneira barata de atender às necessidades nutricionais básicas, dado um conjunto de alimentos. Ele colocou isso como um exercício matemático, não como recomendações alimentares, embora a noção de calcular a nutrição ideal tenha entrado em voga recentemente.

In [1]:
#include <fstream>
#include <string>
#include <vector>

A dieta Stigler determinava que esses mínimos fossem atendidos:

In [2]:
const std::vector<std::pair<std::string, double>> TabelaNutrientes {
    {"Calories (1000s)",3.0},
    {"Protein (grams)",70.0},
    {"Calcium (grams)",0.8},
    {"Iron (mg)",12.0},
    {"Vitamin A (1000 IU)",5.0},
    {"Vitamin B1 (mg)",1.8},
    {"Vitamin B2 (mg)",2.7},
    {"Niacin (mg)",18.0},
    {"Vitamin C (mg)",75.0}
};

O conjunto de alimentos que Stigler avaliou foi um reflexo da época (1944). Os dados nutricionais abaixo são por dólar, não por unidade, portanto, o objetivo é determinar quantos dólares gastar por alimento.

In [3]:
const std::vector<std::pair<std::string, double>> TabelaAlimentos {
    {"Wheat Flour (Enriched) 10 lb.", 36.0},
    {"Macaroni 1 lb.", 14.1},
    {"Wheat Cereal (Enriched) 28 oz.", 24.2},
    {"Corn Flakes 8 oz.", 7.1},
    {"Corn Meal 1 lb.", 4.6},
    {"Hominy Grits 24 oz.", 8.5},
    {"Rice 1 lb.", 7.5},
    {"Rolled Oats 1 lb.", 7.1},
    {"White Bread (Enriched) 1 lb.", 7.9},
    {"Whole Wheat Bread 1 lb.", 9.1},
    {"Rye Bread 1 lb.", 9.1},
    {"Pound Cake 1 lb.", 24.8},
    {"Soda Crackers 1 lb.", 15.1},
    {"Milk 1 qt.", 6.1},
    {"Evaporated Milk (can) 14.5 oz.", 6.7},
    {"Butter 1 lb.", 30.8},
    {"Oleomargarine 1 lb.", 16.1},
    {"Eggs 1 doz.", 32.6},
    {"Cheese (Cheddar) 1 lb.", 24.2},
    {"Cream 1/2 pt.", 14.1},
    {"Peanut Butter 1 lb.", 17.9},
    {"Mayonnaise 1/2 pt.", 16.7},
    {"Crisco 1 lb.", 20.3},
    {"Lard 1 lb.", 9.8},
    {"Sirloin Steak 1 lb.", 39.6},
    {"Round Steak 1 lb.", 36.4},
    {"Rib Roast 1 lb.", 29.2},
    {"Chuck Roast 1 lb.", 22.6},
    {"Plate 1 lb.", 14.6},
    {"Liver (Beef) 1 lb.", 26.8},
    {"Leg of Lamb 1 lb.", 27.6},
    {"Lamb Chops (Rib) 1 lb.", 36.6},
    {"Pork Chops 1 lb.", 30.7},
    {"Pork Loin Roast 1 lb.", 24.2},
    {"Bacon 1 lb.", 25.6},
    {"Ham, smoked 1 lb.", 27.4},
    {"Salt Pork 1 lb.", 16.0},
    {"Roasting Chicken 1 lb.", 30.3},
    {"Veal Cutlets 1 lb.", 42.3},
    {"Salmon, Pink (can) 16 oz.", 13.0},
    {"Apples 1 lb.", 4.4},
    {"Bananas 1 lb.", 6.1},
    {"Lemons 1 doz.", 26.0},
    {"Oranges 1 doz.", 30.9},
    {"Green Beans 1 lb.", 7.1},
    {"Cabbage 1 lb.", 3.7},
    {"Carrots 1 bunch", 4.7},
    {"Celery 1 stalk", 7.3},
    {"Lettuce 1 head", 8.2},
    {"Onions 1 lb.", 3.6},
    {"Potatoes 15 lb.", 34.0},
    {"Spinach 1 lb.", 8.1},
    {"Sweet Potatoes 1 lb.", 5.1},
    {"Peaches (can) No. 2 1/2", 16.8},
    {"Pears (can) No. 2 1/2", 20.4},
    {"Pineapple (can) No. 2 1/2", 21.3},
    {"Asparagus (can) No. 2", 27.7},
    {"Green Beans (can) No. 2", 10},
    {"Pork and Beans (can) 16 oz.", 7.1},
    {"Corn (can) No. 2", 10.4},
    {"Peas (can) No. 2", 13.8},
    {"Tomatoes (can) No. 2", 8.6},
    {"Tomato Soup (can) 10 1/2 oz.", 7.6},
    {"Peaches, Dried 1 lb.", 15.7},
    {"Prunes, Dried 1 lb.", 9.0},
    {"Raisins, Dried 15 oz.", 9.4},
    {"Peas, Dried 1 lb.", 7.9},
    {"Lima Beans, Dried 1 lb.", 8.9},
    {"Navy Beans, Dried 1 lb.", 5.9},
    {"Coffee 1 lb.", 22.4},
    {"Tea 1/4 lb.", 17.4},
    {"Cocoa 8 oz.", 8.6},
    {"Chocolate 8 oz.", 16.2},
    {"Sugar 10 lb.", 51.7},
    {"Corn Syrup 24 oz.", 13.7},
    {"Molasses 18 oz.", 13.6},
    {"Strawberry Preserves 1 lb.", 20.5}
};

Como todos os nutrientes foram normalizados por preço, nosso objetivo é simplesmente minimizar a soma de alimentos.

Em 1944, Stigler calculou a melhor solução que pôde manualmente, observando com tristeza:
> "... não parece haver nenhum método direto de encontrar o mínimo de uma função linear sujeita a condições lineares."

Ele encontrou uma dieta que custava 39,93 dólares por ano, em 1939.

In [4]:
const std::vector<std::vector<double>> TabelaNutricional {
    { 44.7, 1411.0, 2.0, 365.0, 0.0, 55.4, 33.3, 441.0, 0.0},
    { 11.6, 418.0, 0.7, 54.0, 0.0, 3.2, 1.9, 68.0, 0.0},
    { 11.8, 377.0, 14.4, 175.0, 0.0, 14.4, 8.8, 114.0, 0.0},
    { 11.4, 252.0, 0.1, 56.0, 0.0, 13.5, 2.3, 68.0, 0.0},
    { 36.0, 897.0, 1.7, 99.0, 30.9, 17.4, 7.9, 106, 0.0},
    { 28.6, 680.0, 0.8, 80.0, 0.0, 10.6, 1.6, 110.0, 0.0},
    { 21.2, 460.0, 0.6, 41.0, 0.0, 2.0, 4.8, 60.0, 0.0},
    { 25.3, 907.0, 5.1, 341.0, 0.0, 37.1, 8.9, 64.0, 0.0},
    { 15.0, 488.0, 2.5, 115.0, 0.0, 13.8, 8.5, 126.0, 0.0},
    { 12.2, 484.0, 2.7, 125.0, 0.0, 13.9, 6.4, 160.0, 0.0},
    { 12.4, 439.0, 1.1, 82.0, 0.0, 9.9, 3.0, 66.0, 0.0},
    { 8.0, 130.0, 0.4, 31.0, 18.9, 2.8, 3.0, 17.0, 0.0},
    { 12.5, 288.0, 0.5, 50, 0, 0, 0, 0, 0},
    { 11.0, 310.0, 10.5, 18, 16.8, 4, 16, 7, 177},
    { 8.4, 422.0, 15.1, 9, 26, 3, 23.5, 11, 60},
    { 10.8, 9.0, 0.2, 3, 44.2, 0, 0.2, 2, 0},
    { 20.6, 17.0, 0.6, 6, 55.8, 0.2, 0, 0, 0},
    { 2.9, 238.0, 1.0, 52, 18.6, 2.8, 6.5, 1, 0},
    { 7.4, 448.0, 16.4, 19, 28.1, 0.8, 10.3, 4, 0},
    { 3.5, 49.0, 1.7, 3, 16.9, 0.6, 2.5, 0, 17},
    { 15.7, 661.0, 1.0, 48, 0, 9.6, 8.1, 471, 0},
    { 8.6, 18.0, 0.2, 8, 2.7, 0.4, 0.5, 0, 0},
    { 20.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
    { 41.7, 0.0, 0.0, 0.0, 0.2, 0.0, 0.5, 5.0, 0.0},
    { 2.9, 166.0, 0.1, 34, 0.2, 2.1, 2.9, 69, 0},
    { 2.2, 214.0, 0.1, 32, 0.4, 2.5, 2.4, 87, 0},
    { 3.4, 213.0, 0.1, 33, 0, 0, 2, 0, 0},
    { 3.6, 309.0, 0.2, 46, 0.4, 1, 4, 120, 0},
    { 8.5, 404.0, 0.2, 62, 0, 0.9, 0, 0, 0},
    { 2.2, 333.0, 0.2, 139, 169.2, 6.4, 50.8, 316, 525},
    { 3.1, 245.0, 0.1, 20, 0, 2.8, 3.9, 86, 0},
    { 3.3, 140.0, 0.1, 15, 0, 1.7, 2.7, 54, 0},
    { 3.5, 196.0, 0.2, 30, 0, 17.4, 2.7, 60, 0},
    { 4.4, 249.0, 0.3, 37, 0, 18.2, 3.6, 79, 0},
    { 10.4, 152.0, 0.2, 23, 0, 1.8, 1.8, 71, 0},
    { 6.7, 212.0, 0.2, 31, 0, 9.9, 3.3, 50, 0},
    { 18.8, 164.0, 0.1, 26, 0, 1.4, 1.8, 0, 0},
    { 1.8, 184.0, 0.1, 30, 0.1, 0.9, 1.8, 68, 46},
    { 1.7, 156.0, 0.1, 24, 0, 1.4, 2.4, 57, 0},
    { 5.8, 705.0, 6.8, 45, 3.5, 1, 4.9, 209, 0},
    { 5.8, 27.0, 0.5, 36, 7.3, 3.6, 2.7, 5, 544},
    { 4.9, 60.0, 0.4, 30, 17.4, 2.5, 3.5, 28, 498},
    { 1.0, 21.0, 0.5, 14, 0, 0.5, 0, 4, 952},
    { 2.2, 40.0, 1.1, 18, 11.1, 3.6, 1.3, 10, 1998},
    { 2.4, 138.0, 3.7, 80, 69, 4.3, 5.8, 37, 862},
    { 2.6, 125.0, 4.0, 36, 7.2, 9, 4.5, 26, 5369},
    { 2.7, 73.0, 2.8, 43, 188.5, 6.1, 4.3, 89, 608},
    { 0.9, 51.0, 3.0, 23, 0.9, 1.4, 1.4, 9, 313},
    { 0.4, 27.0, 1.1, 22, 112.4, 1.8, 3.4, 11, 449},
    { 5.8, 166.0, 3.8, 59, 16.6, 4.7, 5.9, 21, 1184},
    { 14.3, 336.0, 1.8, 118, 6.7, 29.4, 7.1, 198, 2522},
    { 1.1, 106.0, 0.0, 138, 918.4, 5.7, 13.8, 33, 2755},
    { 9.6, 138.0, 2.7, 54, 290.7, 8.4, 5.4, 83, 1912},
    { 3.7, 20.0, 0.4, 10, 21.5, 0.5, 1, 31, 196},
    { 3.0, 8.0, 0.3, 8, 0.8, 0.8, 0.8, 5, 81},
    { 2.4, 16.0, 0.4, 8, 2, 2.8, 0.8, 7, 399},
    { 0.4, 33.0, 0.3, 12, 16.3, 1.4, 2.1, 17, 272},
    { 1.0, 54.0, 2.0, 65, 53.9, 1.6, 4.3, 32, 431},
    { 7.5, 364.0, 4.0, 134, 3.5, 8.3, 7.7, 56, 0},
    { 5.2, 136.0, 0.2, 16, 12, 1.6, 2.7, 42, 218},
    { 2.3, 136.0, 0.6, 45, 34.9, 4.9, 2.5, 37, 370},
    { 1.3, 63.0, 0.7, 38, 53.2, 3.4, 2.5, 36, 1253},
    { 1.6, 71.0, 0.6, 43, 57.9, 3.5, 2.4, 67, 862},
    { 8.5, 87.0, 1.7, 173, 86.8, 1.2, 4.3, 55, 57},
    { 12.8, 99.0, 2.5, 154, 85.7, 3.9, 4.3, 65, 257},
    { 13.5, 104.0, 2.5, 136, 4.5, 6.3, 1.4, 24, 136},
    { 20.0, 1367.0, 4.2, 345, 2.9, 28.7, 18.4, 162, 0},
    { 17.4, 1055.0, 3.7, 459, 5.1, 26.9, 38.2, 93, 0},
    { 26.9, 1691.0, 11.4, 792, 0, 38.4, 24.6, 217, 0},
    { 0.0, 0.0, 0.0, 0.0, 0.0, 4.0, 5.1, 50.0, 0.0},
    { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.3, 42.0, 0.0},
    { 8.7, 237.0, 3.0, 72, 0, 2, 11.9, 40, 0},
    { 8.0, 77.0, 1.3, 39, 0, 0.9, 3.4, 14, 0},
    { 34.9, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
    { 14.7, 0.0, 0.5, 74.0, 0.0, 0.0, 0.0, 5.0, 0.0},
    { 9.0, 0.0, 10.3, 244.0, 0.0, 1.9, 7.5, 146.0, 0.0},
    { 6.4, 11.0, 0.4, 7.0, 0.2, 0.2, 0.4, 3.0, 0.0}
};

Em 1947, Jack Laderman usou o método simplex (então, uma invenção recente!) Para determinar a solução ideal. Foram necessários 120 dias-homem de nove funcionários em calculadoras de mesa para chegar à resposta.

2. Crie as variáveis e defina o objetivo

- O método MakeNumVar cria uma variável, alimento[i], para cada linha da tabela de alimentos. Como mencionado anteriormente, os dados nutricionais são por dólar; portanto, alimento[i] é a quantidade de dinheiro para gastar em cada alimento i.

- A função objetivo é o custo total dos alimentos, que é a soma das quantidades de alimentos em alimento[i].

- O método SetCoefficient define os coeficientes da função objetivo, que são todos 1 neste caso. Finalmente, o SetMinimization declara que este é um problema de minimização.

3. Definir as restrições

- As restrições para a dieta Stigler exigem que a quantidade total de nutrientes fornecidos por todos os alimentos seja pelo menos o requisito mínimo para cada nutriente. Em seguida, escrevemos essas restrições como desigualdades envolvendo os dados e nutrientes das matrizes e as variáveis alimento[i].

- Primeiro, a quantidade de nutriente i fornecida pelos alimentos j por dólar é definida em restricoesNutricionais[i]. Considerando que a quantidade de dinheiro para ser gasto com o alimento j é  alimento[j], a quantidade de nutriente fornecida pelo alimento j é $alimento[j]\times TabelaNutricional[j][i]$

- Finalmente, como o requisito mínimo para o nutriente i está em TabelaNutrientes[i], podemos escrever a restrição i da seguinte maneira:
\begin{equation}
\sum_{j} TabelaNutricional[j][i] \times TabelaAlimentos[j] \ge TabelaNutrientes[i]
\end{equation}

- cria uma restrição na qual uma combinação linear das variáveis alimento [j] (definida a seguir) é maior ou igual aos nutrientes [i]. Os coeficientes da expressão linear são definidos pelo método SetCoefficient.

In [5]:
#include "setup.h"
#include "ortools/linear_solver/linear_solver.h"
#include "ortools/linear_solver/linear_solver.pb.h"

using namespace operations_research;

In [6]:
{
    // 1. Escolha o solver.
    MPSolver solver("DietaStigler", MPSolver::GLOP_LINEAR_PROGRAMMING);

    // 2. Crie as variáveis:
    std::vector<MPVariable*> alimento(TabelaAlimentos.size());
    for (int i=0; i < alimento.size(); i++) 
        alimento[i] = solver.MakeNumVar(0.0, 1, TabelaAlimentos[i].first);

    // 3. Defina as restrições:
    std::vector<MPConstraint*> restricoesNutricionais(TabelaNutrientes.size());
    for (int i=0; i < TabelaNutrientes.size(); i++) {
        restricoesNutricionais[i] = solver.MakeRowConstraint(TabelaNutrientes[i].second, solver.infinity());
        for (int j=0; j < alimento.size(); j++)
            restricoesNutricionais[i]->SetCoefficient(alimento[j], TabelaNutricional[j][i]);
    }

    // 4. Defina a função objetivo: minimize a soma do preço normalizado dos alimentos.
    auto const objective = solver.MutableObjective();
    for(int i=0; i < alimento.size(); i++) 
        objective->SetCoefficient(alimento[i], 1);
    objective->SetMinimization();
    
    // 5. Execute o solver e exiba os resultados.
    auto status = solver.Solve();
    
    if (status == MPSolver::OPTIMAL) {
        // Display the amounts (in dollars) to purchase of each food.
        auto price = 0.0;
        
        std::vector<double> nutrients(TabelaNutrientes.size());
        for (int i=0; i < TabelaAlimentos.size(); i++){
            price += alimento[i]->solution_value();
            
            for (int j = 0; j < TabelaNutrientes.size(); j++)
                nutrients[j] += TabelaNutricional[i][j] * alimento[i]->solution_value();
            
            if (alimento[i]->solution_value() > 0.0)
                std::cout << TabelaAlimentos[i].first << " = " << alimento[i]->solution_value() << std::endl;
        }
        std::cout << "Optimal annual price: " << (365.0 * price) << std::endl;
    } 
    else if (status == MPSolver::FEASIBLE) // No optimal solution was found.
      std::cout << "A potentially suboptimal solution was found.";
    else
      std::cout << "The solver could not solve the problem.";
    
}

Wheat Flour (Enriched) 10 lb. = 0.0295191
Liver (Beef) 1 lb. = 0.00189256
Cabbage 1 lb. = 0.0112144
Spinach 1 lb. = 0.00500766
Navy Beans, Dried 1 lb. = 0.0610286
Optimal annual price: 39.6617


Dieta Ótima:
- Wheat Flour (Enriched) = 0.029519
- Liver (Beef) = 0.001893
- Cabbage = 0.011214
- Spinach = 0.005008
- Navy Beans, Dried = 0.061029

Optimal annual price: $39.66